components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java

```diff
@@ -143,167 +145,260 @@ protected void doStop() throws Exception {
+
+        final HeaderFilterStrategy headerFilterStrategy = configuration.getHeaderFilterStrategy();
+
+        final String key = entry.getKey();
+        final Object value = entry.getValue();
+
+        if (shouldBeFiltered(key, value, exchange, headerFilterStrategy)) {
+            final KafkaHeaderSerializer headerSerializer = configuration.getHeaderSerializer();
+            final byte[] headerValue = headerSerializer.serialize(key, value);
+
+            if (headerValue == null) {
+                return null;
+            }
+            return new RecordHeader(key, headerValue);
+        }
+
+        return null;
     }
 
     @Override
-    @SuppressWarnings({ "unchecked", "rawtypes" })
     // Camel calls this method if the endpoint isSynchronous(), as the
     // KafkaEndpoint creates a SynchronousDelegateProducer for it
     public void process(Exchange exchange) throws Exception {
-        Iterator<KeyValueHolder<Object, ProducerRecord>> c = createRecorder(exchange);
+
         List<KeyValueHolder<Object, Future<RecordMetadata>>> futures = new LinkedList<>();
-        List<RecordMetadata> recordMetadatas = new ArrayList<>();
+        List<RecordMetadata> recordMetadata = new ArrayList<>();
 
-        if (endpoint.getConfiguration().isRecordMetadata()) {
-            exchange.getMessage().setHeader(KafkaConstants.KAFKA_RECORDMETA, recordMetadatas);
+        if (configuration.isRecordMetadata()) {
+            exchange.getMessage().setHeader(KafkaConstants.KAFKA_RECORDMETA, recordMetadata);
         }
 
-        while (c.hasNext()) {
-            KeyValueHolder<Object, ProducerRecord> exrec = c.next();
-            ProducerRecord rec = exrec.getValue();
-            if (LOG.isDebugEnabled()) {
-                LOG.debug("Sending message to topic: {}, partition: {}, key: {}", rec.topic(), rec.partition(), rec.key());
+        // is the message body a list or something that contains multiple values
+        Message message = exchange.getIn();
+        if (isIterable(message.getBody())) {
+            Iterator<KeyValueHolder<Object, ProducerRecord<Object, Object>>> recordIterable
+                    = createRecordIterable(exchange, message);
+
+            while (recordIterable.hasNext()) {
+                KeyValueHolder<Object, ProducerRecord<Object, Object>> exchangeRecord = recordIterable.next();
+                ProducerRecord<Object, Object> rec = exchangeRecord.getValue();
+
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Sending message to topic: {}, partition: {}, key: {}", rec.topic(), rec.partition(), rec.key());
+                }
+
+                futures.add(new KeyValueHolder<>(exchangeRecord.getKey(), kafkaProducer.send(rec)));
             }
-            futures.add(new KeyValueHolder(exrec.getKey(), kafkaProducer.send(rec)));
+        } else {
+            final KeyValueHolder<Object, ProducerRecord<Object, Object>> kvHolder = createKeyValueHolder(exchange, message);
```

Comment:
Since the message body in most use-cases is not iterable, eg send 1 message, maybe this can be optimized to not use futures and creating key value holders etc.

Reply:
+1. I think I can do it like I am doing on the `process(Exchange exchange, AsyncCallback callback)` one and create the producer record directly.